# Automobile Price Prediction

In [1]:
import pandas as pd
import numpy as np

In [2]:
car = pd.read_csv('car.csv')

In [3]:
car.head()

name   company  year          Price  \
0    Hyundai Santro Xing XO eRLX Euro III   Hyundai  2007         80,000   
1                 Mahindra Jeep CL550 MDI  Mahindra  2006       4,25,000   
2              Maruti Suzuki Alto 800 Vxi    Maruti  2018  Ask For Price   
3  Hyundai Grand i10 Magna 1.2 Kappa VTVT   Hyundai  2014       3,25,000   
4        Ford EcoSport Titanium 1.5L TDCi      Ford  2014       5,75,000   

   kms_driven fuel_type  
0  45,000 kms    Petrol  
1      40 kms    Diesel  
2  22,000 kms    Petrol  
3  28,000 kms    Petrol  
4  36,000 kms    Diesel

In [4]:
car.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 892 entries, 0 to 891
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        892 non-null    object
 1   company     892 non-null    object
 2   year        892 non-null    object
 3   Price       892 non-null    object
 4   kms_driven  840 non-null    object
 5   fuel_type   837 non-null    object
dtypes: object(6)
memory usage: 21.0+ KB


In [5]:
car.shape

(892, 6)

In [6]:
car['year'].tail(5)

887    zest
888    2018
889    2013
890    2014
891    2014
Name: year, dtype: object

#### DATA CLEANING

1. We can see that in car.csv file there are some null values in kms_driven and fuel_type
2. We can see that name,company,year,Price,kms_driven,fuel_type all are datatype object(string) But We know that year,price,kms_driven are int values
3. In Price column there are comma(,) in between numbers.
4. In year column there is random string instead of year (887)

### Year

In [7]:
car['year'].str.isnumeric() 
# it will give Bullian ans False where value is not numeric

0       True
1       True
2       True
3       True
4       True
       ...  
887    False
888     True
889     True
890     True
891     True
Name: year, Length: 892, dtype: bool

In [8]:
car = car[car['year'].str.isnumeric()] #car[year] == isnumeriac

In [9]:
car.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 842 entries, 0 to 891
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        842 non-null    object
 1   company     842 non-null    object
 2   year        842 non-null    object
 3   Price       842 non-null    object
 4   kms_driven  840 non-null    object
 5   fuel_type   837 non-null    object
dtypes: object(6)
memory usage: 26.3+ KB


In [10]:
car['year'] = car['year'].astype(int)

In [11]:
car.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 842 entries, 0 to 891
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        842 non-null    object
 1   company     842 non-null    object
 2   year        842 non-null    int32 
 3   Price       842 non-null    object
 4   kms_driven  840 non-null    object
 5   fuel_type   837 non-null    object
dtypes: int32(1), object(5)
memory usage: 26.3+ KB


### Price

In [12]:
car = car[car['Price']!='Ask For Price'] #(car[price] != ask for price) == True
car.head()

name   company  year     Price  \
0    Hyundai Santro Xing XO eRLX Euro III   Hyundai  2007    80,000   
1                 Mahindra Jeep CL550 MDI  Mahindra  2006  4,25,000   
3  Hyundai Grand i10 Magna 1.2 Kappa VTVT   Hyundai  2014  3,25,000   
4        Ford EcoSport Titanium 1.5L TDCi      Ford  2014  5,75,000   
6                               Ford Figo      Ford  2012  1,75,000   

   kms_driven fuel_type  
0  45,000 kms    Petrol  
1      40 kms    Diesel  
3  28,000 kms    Petrol  
4  36,000 kms    Diesel  
6  41,000 kms    Diesel

In [13]:
car['Price'] = car['Price'].str.replace(',','').astype(int)  #replace , with ""
car.head()

name   company  year   Price  kms_driven  \
0    Hyundai Santro Xing XO eRLX Euro III   Hyundai  2007   80000  45,000 kms   
1                 Mahindra Jeep CL550 MDI  Mahindra  2006  425000      40 kms   
3  Hyundai Grand i10 Magna 1.2 Kappa VTVT   Hyundai  2014  325000  28,000 kms   
4        Ford EcoSport Titanium 1.5L TDCi      Ford  2014  575000  36,000 kms   
6                               Ford Figo      Ford  2012  175000  41,000 kms   

  fuel_type  
0    Petrol  
1    Diesel  
3    Petrol  
4    Diesel  
6    Diesel

In [14]:
car.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 819 entries, 0 to 891
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        819 non-null    object
 1   company     819 non-null    object
 2   year        819 non-null    int32 
 3   Price       819 non-null    int32 
 4   kms_driven  819 non-null    object
 5   fuel_type   816 non-null    object
dtypes: int32(2), object(4)
memory usage: 25.6+ KB


### Kms_driven

In [15]:
car['kms_driven'].str.split()

0        [45,000, kms]
1            [40, kms]
3        [28,000, kms]
4        [36,000, kms]
6        [41,000, kms]
            ...       
886    [1,32,000, kms]
888      [27,000, kms]
889      [40,000, kms]
890           [Petrol]
891           [Petrol]
Name: kms_driven, Length: 819, dtype: object

In [16]:
car['kms_driven'].str.split().str.get(0)

0        45,000
1            40
3        28,000
4        36,000
6        41,000
         ...   
886    1,32,000
888      27,000
889      40,000
890      Petrol
891      Petrol
Name: kms_driven, Length: 819, dtype: object

In [17]:
car['kms_driven'].str.split().str.get(0).str.replace(',','')

0       45000
1          40
3       28000
4       36000
6       41000
        ...  
886    132000
888     27000
889     40000
890    Petrol
891    Petrol
Name: kms_driven, Length: 819, dtype: object

In [18]:
car['kms_driven'] = car['kms_driven'].str.split().str.get(0).str.replace(',','')

In [19]:
car = car[car['kms_driven'].str.isnumeric()]

In [20]:
car['kms_driven'] = car['kms_driven'].astype(int)

In [21]:
car.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 817 entries, 0 to 889
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        817 non-null    object
 1   company     817 non-null    object
 2   year        817 non-null    int32 
 3   Price       817 non-null    int32 
 4   kms_driven  817 non-null    int32 
 5   fuel_type   816 non-null    object
dtypes: int32(3), object(3)
memory usage: 25.5+ KB


In [22]:
~car['fuel_type'].isna()

0      True
1      True
3      True
4      True
6      True
       ... 
883    True
885    True
886    True
888    True
889    True
Name: fuel_type, Length: 817, dtype: bool

In [23]:
car = car[~car['fuel_type'].isna()] 

In [24]:
car.shape

(816, 6)

In [25]:
car.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 816 entries, 0 to 889
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        816 non-null    object
 1   company     816 non-null    object
 2   year        816 non-null    int32 
 3   Price       816 non-null    int32 
 4   kms_driven  816 non-null    int32 
 5   fuel_type   816 non-null    object
dtypes: int32(3), object(3)
memory usage: 25.5+ KB


In [26]:
car.head()

name   company  year   Price  kms_driven  \
0    Hyundai Santro Xing XO eRLX Euro III   Hyundai  2007   80000       45000   
1                 Mahindra Jeep CL550 MDI  Mahindra  2006  425000          40   
3  Hyundai Grand i10 Magna 1.2 Kappa VTVT   Hyundai  2014  325000       28000   
4        Ford EcoSport Titanium 1.5L TDCi      Ford  2014  575000       36000   
6                               Ford Figo      Ford  2012  175000       41000   

  fuel_type  
0    Petrol  
1    Diesel  
3    Petrol  
4    Diesel  
6    Diesel

In [27]:
cars_name = set(car['name'])

In [28]:
len(cars_name)

463

In [29]:
car['name'] = car['name'].str.split().str.slice(0,3).str.join(' ')

In [30]:
car.head()

name   company  year   Price  kms_driven fuel_type
0     Hyundai Santro Xing   Hyundai  2007   80000       45000    Petrol
1     Mahindra Jeep CL550  Mahindra  2006  425000          40    Diesel
3       Hyundai Grand i10   Hyundai  2014  325000       28000    Petrol
4  Ford EcoSport Titanium      Ford  2014  575000       36000    Diesel
6               Ford Figo      Ford  2012  175000       41000    Diesel

In [31]:
cars__name = set(car['name'])
len(cars__name)

254

In [32]:
car

name   company  year   Price  kms_driven fuel_type
0       Hyundai Santro Xing   Hyundai  2007   80000       45000    Petrol
1       Mahindra Jeep CL550  Mahindra  2006  425000          40    Diesel
3         Hyundai Grand i10   Hyundai  2014  325000       28000    Petrol
4    Ford EcoSport Titanium      Ford  2014  575000       36000    Diesel
6                 Ford Figo      Ford  2012  175000       41000    Diesel
..                      ...       ...   ...     ...         ...       ...
883      Maruti Suzuki Ritz    Maruti  2011  270000       50000    Petrol
885          Tata Indica V2      Tata  2009  110000       30000    Diesel
886    Toyota Corolla Altis    Toyota  2009  300000      132000    Petrol
888            Tata Zest XM      Tata  2018  260000       27000    Diesel
889      Mahindra Quanto C8  Mahindra  2013  390000       40000    Diesel

[816 rows x 6 columns]

### Reset indexing

In [33]:
car = car.reset_index(drop=True)

In [34]:
car

name   company  year   Price  kms_driven fuel_type
0       Hyundai Santro Xing   Hyundai  2007   80000       45000    Petrol
1       Mahindra Jeep CL550  Mahindra  2006  425000          40    Diesel
2         Hyundai Grand i10   Hyundai  2014  325000       28000    Petrol
3    Ford EcoSport Titanium      Ford  2014  575000       36000    Diesel
4                 Ford Figo      Ford  2012  175000       41000    Diesel
..                      ...       ...   ...     ...         ...       ...
811      Maruti Suzuki Ritz    Maruti  2011  270000       50000    Petrol
812          Tata Indica V2      Tata  2009  110000       30000    Diesel
813    Toyota Corolla Altis    Toyota  2009  300000      132000    Petrol
814            Tata Zest XM      Tata  2018  260000       27000    Diesel
815      Mahindra Quanto C8  Mahindra  2013  390000       40000    Diesel

[816 rows x 6 columns]

In [35]:
car.to_csv('Cleaned_Car_data.csv')

In [36]:
car.describe()

year         Price     kms_driven
count   816.000000  8.160000e+02     816.000000
mean   2012.444853  4.117176e+05   46275.531863
std       4.002992  4.751844e+05   34297.428044
min    1995.000000  3.000000e+04       0.000000
25%    2010.000000  1.750000e+05   27000.000000
50%    2013.000000  2.999990e+05   41000.000000
75%    2015.000000  4.912500e+05   56818.500000
max    2019.000000  8.500003e+06  400000.000000

In [37]:
car[car['Price']>6000000]  #outlier

name   company  year    Price  kms_driven fuel_type
534  Mahindra XUV500 W6  Mahindra  2014  8500003       45000    Diesel

In [38]:
car = car[car['Price']<6000000]

In [39]:
car.describe()

year         Price     kms_driven
count   815.000000  8.150000e+02     815.000000
mean   2012.442945  4.017933e+05   46277.096933
std       4.005079  3.815888e+05   34318.459638
min    1995.000000  3.000000e+04       0.000000
25%    2010.000000  1.750000e+05   27000.000000
50%    2013.000000  2.999990e+05   41000.000000
75%    2015.000000  4.900000e+05   56879.000000
max    2019.000000  3.100000e+06  400000.000000

In [40]:
set(car['company'])

{'Audi',
 'BMW',
 'Chevrolet',
 'Datsun',
 'Fiat',
 'Force',
 'Ford',
 'Hindustan',
 'Honda',
 'Hyundai',
 'Jaguar',
 'Jeep',
 'Land',
 'Mahindra',
 'Maruti',
 'Mercedes',
 'Mini',
 'Mitsubishi',
 'Nissan',
 'Renault',
 'Skoda',
 'Tata',
 'Toyota',
 'Volkswagen',
 'Volvo'}

In [41]:
len(set(car['company']))

25

In [42]:
car.head()

name   company  year   Price  kms_driven fuel_type
0     Hyundai Santro Xing   Hyundai  2007   80000       45000    Petrol
1     Mahindra Jeep CL550  Mahindra  2006  425000          40    Diesel
2       Hyundai Grand i10   Hyundai  2014  325000       28000    Petrol
3  Ford EcoSport Titanium      Ford  2014  575000       36000    Diesel
4               Ford Figo      Ford  2012  175000       41000    Diesel

In [43]:
X = car[['name','company','year','kms_driven','fuel_type']]
y = car['Price']

In [44]:
X

name   company  year  kms_driven fuel_type
0       Hyundai Santro Xing   Hyundai  2007       45000    Petrol
1       Mahindra Jeep CL550  Mahindra  2006          40    Diesel
2         Hyundai Grand i10   Hyundai  2014       28000    Petrol
3    Ford EcoSport Titanium      Ford  2014       36000    Diesel
4                 Ford Figo      Ford  2012       41000    Diesel
..                      ...       ...   ...         ...       ...
811      Maruti Suzuki Ritz    Maruti  2011       50000    Petrol
812          Tata Indica V2      Tata  2009       30000    Diesel
813    Toyota Corolla Altis    Toyota  2009      132000    Petrol
814            Tata Zest XM      Tata  2018       27000    Diesel
815      Mahindra Quanto C8  Mahindra  2013       40000    Diesel

[815 rows x 5 columns]

In [45]:
y

0       80000
1      425000
2      325000
3      575000
4      175000
        ...  
811    270000
812    110000
813    300000
814    260000
815    390000
Name: Price, Length: 815, dtype: int32

## Train Test Split

In [46]:
from sklearn.model_selection import train_test_split

In [47]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=655)

In [48]:
len(X_train), len(X_test), len(y_train), len(y_test)

(652, 163, 652, 163)

In [49]:
X_train.head()

name   company  year  kms_driven fuel_type
420      Honda Amaze 1.5     Honda  2016      160000    Diesel
722      Ford Fusion 1.4      Ford  2007       85455    Diesel
150   Maruti Suzuki Alto    Maruti  2018        3000    Petrol
347      Mahindra XUV500  Mahindra  2015       15000    Diesel
100  Maruti Suzuki Swift    Maruti  2013       13349    Petrol

## Encoding

In [50]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import make_column_transformer

In [51]:
ohe = OneHotEncoder()
ohe.fit(X[['name','company','fuel_type']])

OneHotEncoder()

In [52]:
ohe.categories_

[array(['Audi A3 Cabriolet', 'Audi A4 1.8', 'Audi A4 2.0', 'Audi A6 2.0',
        'Audi A8', 'Audi Q3 2.0', 'Audi Q5 2.0', 'Audi Q7', 'BMW 3 Series',
        'BMW 5 Series', 'BMW 7 Series', 'BMW X1', 'BMW X1 sDrive20d',
        'BMW X1 xDrive20d', 'Chevrolet Beat', 'Chevrolet Beat Diesel',
        'Chevrolet Beat LS', 'Chevrolet Beat LT', 'Chevrolet Beat PS',
        'Chevrolet Cruze LTZ', 'Chevrolet Enjoy', 'Chevrolet Enjoy 1.4',
        'Chevrolet Sail 1.2', 'Chevrolet Sail UVA', 'Chevrolet Spark',
        'Chevrolet Spark 1.0', 'Chevrolet Spark LS', 'Chevrolet Spark LT',
        'Chevrolet Tavera LS', 'Chevrolet Tavera Neo', 'Datsun GO T',
        'Datsun Go Plus', 'Datsun Redi GO', 'Fiat Linea Emotion',
        'Fiat Petra ELX', 'Fiat Punto Emotion', 'Force Motors Force',
        'Force Motors One', 'Ford EcoSport', 'Ford EcoSport Ambiente',
        'Ford EcoSport Titanium', 'Ford EcoSport Trend',
        'Ford Endeavor 4x4', 'Ford Fiesta', 'Ford Fiesta SXi', 'Ford Figo',
        '

In [53]:
column_trans = make_column_transformer((OneHotEncoder(categories=ohe.categories_),['name','company','fuel_type']), 
                        remainder='passthrough')

In [54]:
from sklearn.linear_model import LinearRegression

In [55]:
lr = LinearRegression()

In [56]:
from sklearn.pipeline import make_pipeline

In [57]:
pipe = make_pipeline(column_trans, lr)

In [58]:
pipe.fit(X_train, y_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(categories=[array(['Audi A3 Cabriolet', 'Audi A4 1.8', 'Audi A4 2.0', 'Audi A6 2.0',
       'Audi A8', 'Audi Q3 2.0', 'Audi Q5 2.0', 'Audi Q7', 'BMW 3 Series',
       'BMW 5 Series', 'BMW 7 Series', 'BMW X1', 'BMW X1 sDrive20d',
       'BMW X1 xDrive20d', 'Chevrolet Beat', 'Chevrolet Beat...
                                                                            array(['Audi', 'BMW', 'Chevrolet', 'Datsun', 'Fiat', 'Force', 'Ford',
       'Hindustan', 'Honda', 'Hyundai', 'Jaguar', 'Jeep', 'Land',
       'Mahindra', 'Maruti', 'Mercedes', 'Mini', 'Mitsubishi', 'Nissan',
       'Renault', 'Skoda', 'Tata', 'Toyota', 'Volkswagen', 'Volvo'],
      dtype=object),
                                                                            array(['Diesel', 'LPG', 'Pe

In [59]:
y_pred = pipe.predict(X_test)

In [60]:
y_pred

array([ 4.92410552e+05,  2.00230274e+05,  1.89132267e+06,  2.25906953e+05,
        3.55597350e+05,  1.35058575e+05,  6.25351764e+05,  6.67114610e+05,
        5.20870890e+05,  3.96656082e+05,  1.26489044e+06,  7.87773961e+04,
        1.31756830e+05,  2.30628661e+05,  1.88561989e+05,  2.81691172e+05,
        5.78778711e+05,  3.02437524e+05,  4.27043188e+05,  5.61479324e+05,
        4.38347224e+05,  3.05976490e+05,  2.34777860e+05,  1.49452971e+05,
        4.36983769e+05,  4.54273593e+05,  2.99923307e+05,  5.11911894e+05,
        2.81060436e+05,  2.92344447e+05,  2.29296166e+05,  1.70244602e+05,
        2.88776885e+05,  3.67156836e+05,  1.55020578e+05,  5.42376925e+05,
       -1.62733681e+02,  2.63526117e+05,  3.37596007e+05,  1.59571917e+05,
        1.89144245e+06,  6.88154277e+04,  3.74962830e+05,  2.59168757e+05,
        1.71872691e+05,  6.69591079e+05,  2.41805477e+05,  3.80544382e+05,
        1.58490653e+05, -5.99710088e+03,  1.07690032e+06,  1.91641203e+05,
        1.52474467e+06,  

In [61]:
from sklearn.metrics import r2_score

In [62]:
r2_score(y_test, y_pred)

0.7406944031903535

In [63]:
pipe.predict(pd.DataFrame(columns=X_test.columns, data=np.array(['Maruti Suzuki Ritz','Maruti',2011,50000,'Petrol']).reshape(1,5)))

array([244553.38758886])

In [64]:
import pickle

In [65]:
pickle.dump(pipe, open('LinearRegressionModel.pkl','wb'))